server/ComputeSuperpixelFeatures/ComputeSuperpixelFeatures.py

```diff
@@ -0,0 +1,387 @@
+
+    slide_pixel_array = np.asarray([features
+                                    for cdata, fdata in tile_result_list
+                                    for features in fdata])
+
+    patch_detection_time = time.time() - start_time
+
+    print 'Number of image patches = ', len(slide_centers)
+    print "Patch detection time taken = %s" % cli_utils.disp_time_hms(patch_detection_time)
+
+    #
+    # Fit to autoencoder model to extract the encoded features
+    #
+    print('\n>> Training model to extract low dimensional features ...\n')
+
+    start_time = time.time()
+
+    encoded_features = get_encoded_features(slide_pixel_array, args.inputModelFile, args)
+
+    encoding_time = time.time() - start_time
+
+    print 'Encoding time = %s' % cli_utils.disp_time_hms(encoding_time)
+
+    #
+    # Get slide index, slide patch index, slide name, mean, standard deviation, and x, y centers
+    #
+    slide_patch_num = len(slide_centers)
+    slide_num = 1
+
+    # get slide names
+    slide_names = []
+    base = os.path.basename(args.inputImageFile)
+    slide_name = os.path.splitext(base)[0]
+    slide_names = np.append(slide_names, slide_name)
+
+    # get slide index for each image patch
+    slide_index = np.zeros((slide_patch_num, 1), dtype=np.int)
+    for i in range(slide_patch_num):
+        slide_index[i, 0] = i
+
+    # get index of the first patch for each slide
+    slide_patch_index = np.zeros((slide_num, 1), dtype=np.int)
+    for i in range(slide_num):
+        slide_patch_index[i, 0] = 0
+
+    # get mean and standard deviation for each feature
+    slide_feature_num = encoded_features.shape[1]
+    slide_feature_mean = np.zeros((slide_feature_num, 1), dtype=np.float32)
+    slide_feature_stddev = np.zeros((slide_feature_num, 1), dtype=np.float32)
+    slide_feature_mean[:, 0] = np.mean(encoded_features, axis=0)
+    slide_feature_stddev[:, 0] = np.std(encoded_features, axis=0)
+
+    # get x, y centers for each image patch
+    slide_center_x_list = np.zeros((slide_patch_num, 1), dtype=np.float32)
+    slide_center_y_list = np.zeros((slide_patch_num, 1), dtype=np.float32)
+    slide_center_x_list[:, 0] = slide_centers[:, 0]
+    slide_center_y_list[:, 0] = slide_centers[:, 1]
+
+    #
+    # Create Feature file
+    #
+    print('>> Writing H5 feature file')
+
+    if feature_file_format == '.h5':
+
+        output = h5py.File(args.outputSuperpixelFeatureFile, 'w')
+        output.create_dataset('slides', data=slide_names)
+        output.create_dataset('slideIdx', data=slide_index)
+        output.create_dataset('dataIdx', data=slide_patch_index)
+        output.create_dataset('mean', data=slide_feature_mean)
+        output.create_dataset('std_dev', data=slide_feature_stddev)
+        output.create_dataset('features', data=encoded_features)
+        output.create_dataset('x_centroid', data=slide_center_x_list)
+        output.create_dataset('y_centroid', data=slide_center_y_list)
+        output.close()
```

Comment:
In addition to the center, may be we should also store the patch-size. People may run this CLI with different patch sizes. If that is not stored in the output file format then, it will get lost.

Also, we need make any changes necessary to minimize the effort in loading this file into HistomicsML